### DISCUSSION TITLE
Free RTOS

### DISCUSSION THREAD

**Discussion by RepRapid** [tokio-rs/tokio#4617] [Q&A]
hello, this is more to see if what I'm trying to do can be done in compiling an existing rust application. The application uses Tokio, compiles and runs on linux just fine. The application is small enough to run on smaller less power hungry processor, its going to be for IOT device I have a couple chips in mind, that are 32 bit, and can run FreeRTOS. Can Tokio compile on non linux based Rust?

**Comment by Noah-Kennedy** [comment]
Tokio supports a number of operating systems, but cannot run in a `#![no_std]` environment or on FreeRTOS, as FreeRTOS lacks a lot of the base features required for Tokio to support it.

**Reply by RepRapid** [reply]
I don't know anything about Rust, I do a lot of work with Espressif esp32, Which I know has the FreeRtos option... but I recently saw that they got rust to work with their mcu, Not sure if that would work, its a program that runs on the linux systems. The extent of my Rust knowledge is just compiling it. Im sure I can read through it and rewrite it in C or microPython, But Id rather avoid the learning curve. That is why I'm asking before I even dip my toes into it. Thanks for your response.
Its a simple enough program that would take me a day or two to read through and rewrite, but the MCU already has a TCP stack and what not, and can do async, it has 2 cores and a 3rd ulp, the program is small enough, to fit on it and run with the 32mb version I got , and most likly the 16mb version that I got as well.

**Reply by Darksonn** [reply]
The main limitation in having Tokio support new platforms is the mio crate. You can see which OS mio supports [here][1].

[1]: https://github.com/tokio-rs/mio/#platforms

**Reply by Noah-Kennedy** [reply]
@RepRapid you can get Rust to work on microcontrollers just fine, but just as is the case with C, not all libraries will work, as some require various OS APIs to function.

**Reply by RepRapid** [reply]
I see, thank you for the information. Rust seems interesting, Ive looked into it a little, I think I will have to spend a month or so to get familiar with all the " jargon " and syntax, so that it makes sense to me. But your help has steered me to the right direction, on this specific topic. Thank you.

**Reply by Noah-Kennedy** [reply]
No problem!
